Fixed #30064 -- Added form to validate admin search fields query input.

cansarigol opened the pull request:
Ticket: [#30064](https://code.djangoproject.com/ticket/30064)

charettes commented on django/contrib/admin/views/main.py:
```diff
@@ -35,6 +37,26 @@
+            if var == PAGE_VAR:
+                field = forms.IntegerField()
+            else:
+                field = forms.CharField(validators=[ProhibitNullCharactersValidator])
```
[This validator is added by default, no need to specify it](https://github.com/django/django/blob/ac956dae1d06ce2ebff7a2966bcaf8a5ecdbb861/django/forms/fields.py#L219). We'll probably want to pass `strip=False` as well to preserve backward compatiblity.

Given you have to branch on `var == PAGE_VAR` anyway I don't think there's much value in using a loop here. You could define fields by directly assigning to `self.fields`

```python
def __init__(self, *args, **kwargs):
    self.fields = {
        SEARCH_VAR: forms.CharField(required=False, strip=False, initial=''),
        PAGE_VAR: forms.IntegerField(required=False, min_value=0, initial=0),
        TO_FIELD_VAR: forms.CharField(required=False),
    }
    super(ChangeListForm, self).__init__()
```

The `initial` usage will make the `clean()` below unnecessary.
```

cansarigol replied:
I understood now that ProhibitNullCharactersValidator was appended as CharField's default validator. I missed it. 
your code is very clear and wisely. I applied it

charettes commented on django/contrib/admin/views/main.py:
```diff
@@ -57,16 +79,20 @@ def __init__(self, request, model, list_display, list_display_links,
+            raise forms.ValidationError(change_list_form.errors)
+
+        change_list_form_cleaned = change_list_form.clean()
+        self.page_num = change_list_form_cleaned.get(PAGE_VAR)
```
Simply access `form.cleaned_data` instead.

charettes commented on django/contrib/admin/views/main.py:
```diff
@@ -57,16 +79,20 @@ def __init__(self, request, model, list_display, list_display_links,
-            self.page_num = 0
+        change_list_form = ChangeListForm(request.GET)
+        if not change_list_form.is_valid():
+            raise forms.ValidationError(change_list_form.errors)
```
I think this is problematic as it will result in an unhandled exception instead of displaying an error to the user.

cansarigol replied:
I try to find a sample code about it but I didn't. Can we add an error message into django.contrib.messages with a specific key? we would parse this error message and cover searchbar(search_form.html) input with error field tag.

timgraham replied:
Using messages might work.

timgraham commented on django/contrib/admin/views/main.py:
```diff
@@ -35,6 +36,16 @@
+class ChangeListForm(forms.Form):
+    def __init__(self, *args, **kwargs):
+        super(ChangeListForm, self).__init__(*args, **kwargs)
+        self.fields = {
```
Is there a reason these fields aren't declared on the class?

cansarigol replied:
sorry, I couldn't understand this.

timgraham replied:
Oh, it's because `SEARCH_VAR`, etc. are variables.

charettes replied:
@timgraham I confirm this was my intent to make sure that constants are used as fieldnames.

timgraham commented on django/contrib/admin/views/main.py:
```diff
@@ -35,6 +36,16 @@
+    def __init__(self, *args, **kwargs):
+        super(ChangeListForm, self).__init__(*args, **kwargs)
+        self.fields = {
+            SEARCH_VAR: forms.CharField(required=False, strip=False, initial=''),
```
I'm a bit sketpical that we wouldn't want to to strip the search. Were there failing tests? I think `initial=''` isn't needed because that's the default, isn't it?

cansarigol replied:
strip is @charettes suggestion because of backward compatiblity.
as I can see the initial is None. At the current code, I've set SEARCH_VAR with init to make sure convert this line properly.
`self.query = request.GET.get(SEARCH_VAR, '')`

carltongibson replied:
I'm not quite following the BC concern...

cansarigol replied:
@timgraham I could understand the problem about default value.when I check the error that @carltongibson showed. I've changed these and remove initial.

```
self.page_num = change_list_form.cleaned_data.get(PAGE_VAR) or 0
self.query = change_list_form.cleaned_data.get(SEARCH_VAR) or ''
```

carltongibson commented on django/forms/utils.py:
```diff
@@ -65,6 +65,11 @@ def as_ul(self):
             format_html_join('', '<li>{}{}</li>', self.items())
         )
 
+    def as_div(self):
```
I'm not seeing the motivation for adding this method. 

* Why wouldn't `as_ul()` do? (Or just the existing `__str__()`  implementation.) 
* Given that the `admin/base.html` already wraps in a `ul`, would `as_text()` not do equally? (Are we really expecting multiple errors?)
* Why sure about `<b>` tags? 
* If we want this, tests?

cansarigol replied:
`as_div` is just a suggestion to provide a nicer view for users. 
`as_ul` makes the result view more separated. the inner `<li>` tags don't have error class therefore they seem with `icon-yes.svg`.
your're right, for this case `as_text` is more eligiable if we have just an error. I've changed with `as_text`.

cansarigol commented:
> There's half a dozen failures here. Two causes:
> 
> ```
>   File "django/contrib/admin/views/main.py", line 239, in get_results
>     result_list = paginator.page(self.page_num + 1).object_list
> TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'
> ```
> and
> 
> ```
> File "django/contrib/messages/api.py", line 30, in add_message
>     'You cannot add messages without installing '
> django.contrib.messages.api.MessageFailure: You cannot add messages without installing django.contrib.messages.middleware.MessageMiddleware
> ```

`MessageMiddleware` error comes from `tests.admin_changelist.test_result_list_editable` because we changed the query parameter validation method. I've changed this part, is it ok?
```
def test_result_list_editable(self):
    ...
    with self.assertRaises(MessageFailure):
                m.get_changelist_instance(request)
```

carltongibson commented:
Errr... @cansarigol, hi. — I would have thought the correct response was to make sure `MIDDLEWARE` was right...

Looking [here](https://github.com/django/django/commit/371ece2f0682e51f2f796854d3e091827a7cea63#diff-044a46e68cb93c11c2dc9a7392119915R90), `MessageMiddleware` is a requirement of the admin. 

Initial thought would be to use `override_settings` to correct the failures but it should be there by default...

https://github.com/django/django/blob/75840688f94c022cd2a62143149a30e6d2db8496/tests/runtests.py#L72-L78

... so first question is where's it being taken out? Can you look? Can we adjust that?

cansarigol commented:
I guess `MessageMiddleware` isn't problem because it is already in `MIDDLEWARE` list. But `request_factory` doesn't have a `_messages` property.

cansarigol commented:
hi @carltongibson, I applied changelist test as per fix and moved the page_num invalid query test to admin_filter test. I also separated the first test changing for you can review. I know that this is not what you except.

By the way, I couldn't separate `PAGE_VAR` and `TO_FIELD_VAR`, could you give a suggestion?

carltongibson commented on tests/admin_filters/tests.py:
```diff
@@ -1038,6 +1039,22 @@ def test_filter_with_failing_queryset(self):
+            modeladmin.get_changelist_instance(request)
+            return request
+
+        for case, error in cases:
```
Use `subTest()` here.

cansarigol replied:
ok

carltongibson commented on tests/admin_filters/tests.py:
```diff
@@ -1038,6 +1039,22 @@ def test_filter_with_failing_queryset(self):
 
+    def test_filter_for_change_list_form(self):
+        modeladmin = BookAdmin(Book, site)
+        cases = ({'q': '\x00'}, 'Null characters are not allowed'), \
```
Any reason not to just use a list or explicit tuple here? (The `\` continuation is a little strange.)

cansarigol replied:
I tried to make a tuple which contains tuple items but I've done wrong. I fixed it

carltongibson commented on django/contrib/admin/views/main.py:
```diff
@@ -57,16 +69,19 @@ def __init__(self, request, model, list_display, list_display_links,
+        if not change_list_form.is_valid():
+            messages.error(request, change_list_form.errors.as_text())
+
+        self.page_num = change_list_form.cleaned_data.get(PAGE_VAR) or 0
```
[`get()` takes a default to return as a second parameter](https://docs.python.org/3.7/library/stdtypes.html#dict.get).

cansarigol replied:
this is about page_num param. I couldn't find but the page_num value may be None. So I did like this.
```
>>> x = {'a': 1, 'b': None}
>>>x.get('b', 1)
None
>>>x.get('b', 1) or 1
1
>>>x.get('c', 1)
1
>>> 
```

cansarigol commented:
> Hi @cansarigol
> 
> * Create the form with just `SEARCH_VAR` field for this PR. (Which will be backported as a bug fix.)
> * Then add the other fields and changes in a separate cleanup ticket/PR. (That will target 3.0.)
> 
> Make sense?

Hi, @carltongibson thanks for your review and help. it looks great. 
I pushed two commits. If it's ok, I will separate the second one as a new PR.

timgraham commented:
I don't see this ticket as qualifying for a backport to stable/2.2.x.

carltongibson commented:
It’s a bug fix no? Current major version gets it?

timgraham commented:
Only select bugs per the [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions). This issue doesn't seem critical.

carltongibson commented:
OK, for me, this was a crashing bug from unfiltered user input. On that basis I'd assumed we would port it to pre-release 2.2 (and probably 2.1, since that's still in mainstream support). (2.2 being the new LTS, in thought at least, lending a little extra weight.) 

Happy to defer to your judgement if you say otherwise. (Please confirm, since... — If so, @cansarigol: sorry I led you up the wrong path: we'll take this as a single fix/PR. Sorry.)

timgraham commented:
Part of my rationale is that admin users are generally "trusted." If we were going to backport it, we would have marked it as a release blocker and acted more quickly. The issue has been there since March 2017 and psycopg 2.7.

carltongibson commented:
OK. 👍. (@cansarigol Can you adjust? Sorry again.)

cansarigol commented:
hi, @carltongibson, no problem at all. 👍 I've merged the commits as a PR again.

carltongibson commented on django/contrib/admin/views/main.py:
```diff
@@ -57,6 +67,9 @@ def __init__(self, request, model, list_display, list_display_links,
         # Get search parameters from the query string.
+        change_list_form = ChangeListForm(request.GET)
+        if not change_list_form.is_valid():
+            messages.error(request, change_list_form.errors.as_text())
```
The output here is suboptimal. e.g.`* p * Ensure this value is greater than or equal to 0`

We need to provide custom error messages that make sense when displayed as `messages`. We should maybe add the individual messages, in a loop, rather than the list.

carltongibson commented on tests/admin_filters/tests.py:
```diff
@@ -1038,6 +1039,23 @@ def test_filter_with_failing_queryset(self):
+            ({'q': '\x00'}, 'Null characters are not allowed'),
+        )
+
+        def get_request(kwargs):
```
I'd inline this helper within the subTest()

carltongibson commented on tests/admin_filters/tests.py:
```diff
@@ -1038,6 +1039,23 @@ def test_filter_with_failing_queryset(self):
+
+        for case, error in tests:
+            with self.subTest(case=case):
+                self.assertIn(error, str(next(iter(get_request(case)._messages))))
```
Rather than `str(next(iter(...` maybe: 

```
                messages = [m.message for m in request._messages]
                self.assertEqual(1, len(messages))
                self.assertIn(error, messages[0])
```

cansarigol commented:
Hi, @carltongibson 
 > Can we handle the errors, adding the messages but, maintaining the existing redirect behaviour?
>
I guess we have two options for this:
1- Paginator validator already [checks this case out](https://github.com/django/django/blob/dffa3e1992562ba60512d96d1eb5859ffff2ceb5/django/core/paginator.py#L52). We can carry this error with paginator instance and use in `ChangeList.get_results` 
2- In `ChangeList.get_results` we can validate it again.
Which case makes sense?

>It strikes me that the out of range positive integer is much more likely as input than the -1 case. It would be nice if you could add a decent message in that case. (Can we add an appropriate validator to the page number field?)
>
I agree with you about the positive integer case. But we don't know how many pages we need before `get_results` is called. IMO paginator validator is more convenient than page number field validator. What do you think?

>Also, as per comment inline, we should provide better error messages, and make sure they're displayed nicely.
>
Is this about new clean_up pr about validator messages (`MinValueValidator.message`) or override this messages from `ChangeList`?

carltongibson commented:
Hi @cansarigol. Right, let's get this in...

1. Please strip the form down **just** to the `SEARCH_VAR` field. (`PAGE_VAR` is handled by the paginator, and I don't think we should touch that here. `TO_FIELD_VAR` isn't really subject to user input, so let's address it separately, as and when an issue actually arises.) Rename the form `ChangeListSearchForm`. 
2. Please expose a `search_form_class` attribute on `ChangeList`, setting it to `ChangeListSearchForm`. That way users can override in `get_changelist()` if they need to add an additional validator, say. 
   * If folks need more customization than this, they can override `get_search_results()`, as is already documented. 

This will resolve this ticket and the Admin aspect of https://code.djangoproject.com/ticket/28992
(Super job. Well done.) 

For bonus points, in a separate commit, if you wanted to, you could add tests for the _correction of `PAGE_VAR`_ behaviour (such that this PR as is would have broken those). 

All make sense? Let me know if not!!!

Thank you for your effort here. Good stuff.

cansarigol commented:
Hi @carltongibson, thanks for your help. Could you review first commit?

carltongibson commented on django/contrib/admin/views/main.py:
```diff
@@ -35,10 +37,19 @@
                  list_filter, date_hierarchy, search_fields, list_select_related,
-                 list_per_page, list_max_show_all, list_editable, model_admin, sortable_by):
+                 list_per_page, list_max_show_all, list_editable, model_admin, sortable_by,
+                 search_form_class=ChangeListSearchForm):
```
I meant just as a class attribute, rather than passing it here... Maybe we could do this... (What are your thoughts?)

carltongibson commented:
Follow up in #11700
